File: Sources/HopsCore/Capability.swift
import Foundation

public enum NetworkCapability: String, Codable, Sendable, CaseIterable {
    case disabled
    case outbound
    case loopback
    case full
}

public enum FilesystemCapability: String, Codable, Sendable, CaseIterable {
    case read
    case write
    case execute
}

public struct ResourceLimits: Codable, Sendable, Equatable {
    public var cpus: UInt?
    public var memoryBytes: UInt64?
    public var maxProcesses: UInt?
    
    public init(
        cpus: UInt? = nil,
        memoryBytes: UInt64? = nil,
        maxProcesses: UInt? = nil
    ) {
        self.cpus = cpus
        self.memoryBytes = memoryBytes
        self.maxProcesses = maxProcesses
    }
}

public struct CapabilityGrant: Codable, Sendable, Equatable {
    public var network: NetworkCapability
    public var filesystem: Set<FilesystemCapability>
    public var allowedPaths: Set<String>
    public var deniedPaths: Set<String>
    public var resourceLimits: ResourceLimits
    
    public init(
        network: NetworkCapability = .disabled,
        filesystem: Set<FilesystemCapability> = [],
        allowedPaths: Set<String> = [],
        deniedPaths: Set<String> = [],
        resourceLimits: ResourceLimits = ResourceLimits()
    ) {
        self.network = network
        self.filesystem = filesystem
        self.allowedPaths = allowedPaths
        self.deniedPaths = deniedPaths
        self.resourceLimits = resourceLimits
    }
    
    public static var `default`: CapabilityGrant {
        CapabilityGrant(
            network: .disabled,
            filesystem: [],
            allowedPaths: [],
            deniedPaths: [],
            resourceLimits: ResourceLimits()
        )
    }
}

File: hops-gui/src/models/capability.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Sandbox {
    pub root: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Capabilities {
    pub network: String,
    pub filesystem: FilesystemCapability,
}

impl Default for Capabilities {
    fn default() -> Self {
        Self {
            network: String::from("disabled"),
            filesystem: FilesystemCapability::default(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FilesystemCapability {
    pub read: Vec<String>,
    pub write: Vec<String>,
    pub execute: Vec<String>,
}

impl Default for FilesystemCapability {
    fn default() -> Self {
        Self {
            read: vec![String::from(".")],
            write: vec![],
            execute: vec![String::from(".")],
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Resources {
    pub cpus: u32,
    pub memory: String,
    pub max_processes: u32,
}

impl Default for Resources {
    fn default() -> Self {
        Self {
            cpus: 2,
            memory: String::from("512M"),
            max_processes: 100,
        }
    }
}
